enhancements/microshift/microshift-apiserver-certs.md

```diff
@@ -0,0 +1,355 @@
+the IP will get the internal certificate and wont validate against the external
+CA.
+
+If we were to include the node IP in the external certificate we also get an
+immediate match in the mapping, and the hostname verification also works ok.
+All internal connections (those coming from pods through the service network
+IP) will be broken when trying to validate hostname, as the service IP is
+not included. Even if it was included, validation with the internal CA would
+still fail.
+
+Including the IP in the current situation means choosing what to lose:
+internal or external connectivity. Since internal connectivity is mandatory,
+the only option is to use names for external connectivity. MicroShift will run
+in environments where hostname (or DNS) access may not be guaranteed. Customers
+have given us the requirement of accessing the cluster via IP addresses.
+
+### Using a different IP
+To remove the ambiguity when selecting a certificate based on the destination
+IP, which is the one used to connect to API server, there is a need to have
+more than one network.
+
+The external network is used for external connections, such as `oc` clients.
+The internal network is used for internal connections, such as the ones from
+pods in the cluster.
+
+Choosing the internal network is a complex task, as it can collide with other
+IP ranges coming from DHCP or other environment configurations. In order to
+minimize risks, the API server service IP, which is always the [first one](https://github.com/kubernetes/kubernetes/blob/master/pkg/controlplane/services.go#L47)
+from the service CIDR range, is selected by default.
+
+Since API server listens in [all interfaces](https://github.com/openshift/microshift/blob/0f4e4e8d7cb9946a7af81550eb6a537d0f5b4e15/pkg/controllers/kube-apiserver.go#L185)
+we only need to configure this IP in any interface. To make the solution not
+dependent on CNI an already existing interface may be used, like loopback.
+Upon start, MicroShift API server will configure the service network IP in the
+loopback interface with a 32-bit netmask.
+
+By doing so, the API server service network resolves to a local IP (which
+happens to be the same IP), while the external traffic still uses the external
+IP, which is the node IP. This removes the ambiguity problem and allows the
+external certificate to include the node IP in the SAN. The internal
+certificate remains unchanged, as it already includes the service IP, which is
+now a valid IP within the node.
+
+From a pod's perspective, it is irrelevant whether the IP is local or not. All
+a pod needs is to be able to reach the IP in order to connect to the API
+server. This enables the possibility of adding more nodes to the cluster, but
+keeping the external IP authentication the master node needs to have more than
+one external IP.
+
+### Workflow Description
+Workflow is transparent for the user, as certificates generate themselves upon
+restarting MicroShift.
```

Comment:
Let's use this section to describe how the IP is managed when the system boots and MicroShift start up. The actors may all be software components, but it still helps to explain which of those is doing each thing to make sure we assign the responsibilities correctly.